## ColumnStores

Column stores, also known as columnar databases, are a type of database management system (DBMS) that store data in a column-oriented fashion, as opposed to the traditional row-oriented approach used in relational databases. 

In this lab you will run queries in clickhouse which is a popular column oriented database management system and compare the runtime of those with postgres

In [1]:
import pandas as pd
from sqlalchemy import create_engine, text
from sqlalchemy import Table, Column, Integer, String, Float, DateTime,MetaData, Numeric
from sqlalchemy.dialects.postgresql import TIMESTAMP


## Dataset

The Dataset contains information related to taxi trips in New York City. The following columns are present in the dataset

- **vendorid**: Identifier for the taxi vendor.
- **tpep_pickup_datetime**: Date and time when the trip started.
- **tpep_dropoff_datetime**: Date and time when the trip ended.
- **passenger_count**: Number of passengers in the taxi.
- **trip_distance**: Distance traveled during the trip, likely in miles.
- **ratecodeid**: Rate code for the trip (e.g., standard rate, negotiated rate).
- **store_and_fwd_flag**: Flag indicating whether the trip record was stored in the vehicle's memory before being sent to the vendor.
- **pulocationid**: Location ID for the pickup location.
- **dolocationid**: Location ID for the drop-off location.
- **payment_type**: Payment method used for the trip.
- **fare_amount**: Base fare amount for the trip.
- **extra**: Extra charges, if any, applied to the fare.
- **mta_tax**: Metropolitan Transportation Authority (MTA) tax applied to the fare.
- **tip_amount**: Tip amount given by the passenger.
- **tolls_amount**: Amount of tolls paid during the trip.
- **improvement_surcharge**: Surcharge applied to the fare for improvements (e.g., congestion surcharge).
- **total_amount**: Total fare amount paid by the passenger.
- **congestion_surcharge**: Surcharge applied to the fare due to congestion pricing regulations.


In [2]:
df = pd.read_csv('data/taxi_trip_data_250k.csv')

In [3]:
df.head()

,vendorid,tpep_pickup_datetime,tpep_dropoff_datetime,passenger_count,trip_distance,ratecodeid,store_and_fwd_flag,pulocationid,dolocationid,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge
0,2.0,2019-07-02 20:42:04,2019-07-02 20:44:59,1.0,0.72,1.0,N,90,249,1.0,4.5,0.5,0.5,1.00,0.0,0.3,9.30,2.5
1,1.0,2019-10-09 10:26:41,2019-10-09 11:11:50,1.0,7.50,1.0,N,100,202,1.0,32.0,2.5,0.5,7.06,0.0,0.3,42.36,2.5
2,1.0,2019-06-29 21:05:48,2019-06-29 21:20:01,1.0,1.80,1.0,N,164,161,1.0,11.0,3.0,0.5,2.00,0.0,0.3,16.80,2.5
3,2.0,2020-04-22 23:12:23,2020-04-22 23:21:33,1.0,2.71,1.0,N,142,263,2.0,10.5,0.5,0.5,0.00,0.0,0.3,14.30,2.5
4,1.0,2019-10-25 07:19:54,2019-10-25 07:34:35,1.0,2.80,1.0,N,100,140,1.0,12.0,2.5,0.5,1.00,0.0,0.3,16.30,2.5


In [4]:
df.shape

(250000, 18)

In [5]:
df['tpep_pickup_datetime'] = pd.to_datetime(df['tpep_pickup_datetime']).dt.tz_localize('UTC')
df['tpep_dropoff_datetime'] = pd.to_datetime(df['tpep_dropoff_datetime']).dt.tz_localize('UTC')


In [6]:
df['store_and_fwd_flag'] = df['store_and_fwd_flag'].astype(str)


In [7]:
df = df.fillna(0)

In [8]:
df.isna().sum()

vendorid                 0
tpep_pickup_datetime     0
tpep_dropoff_datetime    0
passenger_count          0
trip_distance            0
ratecodeid               0
store_and_fwd_flag       0
pulocationid             0
dolocationid             0
payment_type             0
fare_amount              0
extra                    0
mta_tax                  0
tip_amount               0
tolls_amount             0
improvement_surcharge    0
total_amount             0
congestion_surcharge     0
dtype: int64

In [9]:
df.dtypes


vendorid                             float64
tpep_pickup_datetime     datetime64[ns, UTC]
tpep_dropoff_datetime    datetime64[ns, UTC]
passenger_count                      float64
trip_distance                        float64
ratecodeid                           float64
store_and_fwd_flag                    object
pulocationid                           int64
dolocationid                           int64
payment_type                         float64
fare_amount                          float64
extra                                float64
mta_tax                              float64
tip_amount                           float64
tolls_amount                         float64
improvement_surcharge                float64
total_amount                         float64
congestion_surcharge                 float64
dtype: object

## Creating connection to Clickhouse and Postgres Server

In [10]:
clickhouse_connection_string = 'clickhouse://admin:click@clickhouse:8123/default'
postgres_connection_string = 'postgresql://admin:password@postgres:5432/clickhouse_pg_db'


In [11]:
clickhouse_engine = create_engine(clickhouse_connection_string)
postgres_engine = create_engine(postgres_connection_string)


## Loading Data to Clickhouse and Postgres DB

In [12]:
def insert_dataframe_to_clickhouse(df, table_name):
    # Truncate the table first
    truncate_query = f'TRUNCATE TABLE IF EXISTS {table_name}'
    client.execute(truncate_query)
    
    data = df.to_dict('records')

    # Insert data into ClickHouse
    client.execute(f'INSERT INTO {table_name} VALUES', data)



In [13]:
from clickhouse_driver import Client

# Define ClickHouse connection parameters
clickhouse_connection_settings = {
    'host': 'clickhouse',
    'port': 9000,
    'user': 'default',
    'password': '',
}

# Initialize ClickHouse client
client = Client(**clickhouse_connection_settings)

In [14]:
client.execute('DROP TABLE IF EXISTS trips_taxi;')
create_table_query ='''
CREATE TABLE trips_taxi
(
    vendorid Float32,
    tpep_pickup_datetime DateTime64(3),
    tpep_dropoff_datetime DateTime64(3),
    passenger_count Float32,
    trip_distance Float32,
    ratecodeid Float32,
    store_and_fwd_flag String,
    pulocationid Int32,
    dolocationid Int32,
    payment_type Float32,
    fare_amount Float32,
    extra Float32,
    mta_tax Float32,
    tip_amount Float32,
    tolls_amount Float32,
    improvement_surcharge Float32,
    total_amount Float32,
    congestion_surcharge Float32
) ENGINE = MergeTree()
ORDER BY (tpep_pickup_datetime, tpep_dropoff_datetime);
'''

# # Execute create table query
client.execute(create_table_query)

[]

In [15]:
insert_dataframe_to_clickhouse(df, 'trips_taxi')

In [16]:
metadata = MetaData()
metadata.create_all(postgres_engine)

In [17]:
import time
from sqlalchemy import create_engine
from clickhouse_sqlalchemy import make_session
from sqlalchemy.orm import sessionmaker

# PostgreSQL connection
PostgresSession = sessionmaker(bind=postgres_engine)
postgres_session = PostgresSession()

# ClickHouse connection
clickhouse_session = make_session(clickhouse_engine)

In [18]:
trips = Table('trips_taxi', metadata,
    Column('vendorid', Float),
    Column('tpep_pickup_datetime', TIMESTAMP(timezone=True)),
    Column('tpep_dropoff_datetime', TIMESTAMP(timezone=True)),
    Column('passenger_count', Float),
    Column('trip_distance', Numeric),
    Column('ratecodeid', Float),
    Column('store_and_fwd_flag', String(1)),
    Column('pulocationid', Integer),
    Column('dolocationid', Integer),
    Column('payment_type', Float),
    Column('fare_amount', Numeric),
    Column('extra', Numeric),
    Column('mta_tax', Numeric),
    Column('tip_amount', Numeric),
    Column('tolls_amount', Numeric),
    Column('improvement_surcharge', Numeric),
    Column('total_amount', Numeric),
    Column('congestion_surcharge', Numeric),
)

In [19]:

# insert the new DataFrame into the empty table
df.to_sql('trips_taxi', con=postgres_engine, if_exists='replace', index=False)


1000

## Functions for running query and comparing runtime

- We have provided some helper functions which you can use to run your queries and compare the runtime.

To run the query in clickhouse and print the dataframe

In [20]:
def run_query(query):
    try:

        # Establish a connection
        with clickhouse_engine.connect() as connection:
            # Define the search query with SQLAlchemy's text construct
            query = text(query)
            # Execute the search query
            result = connection.execute(query)
            # Fetch the result and convert it to a DataFrame
            clickhouse_df = pd.DataFrame(result.fetchall(), columns=result.keys())
            return clickhouse_df

    except Exception as e:
        print("Query execution for ClickHouse failed:", e)
        return 

To compare runtime of postgres and clickhouse queries.

In [21]:
import time
from sqlalchemy import text

def compare_query_runtime(query):
    query = text(query)
    # Execute and time the PostgreSQL query
    try:
        start_time = time.time()
        result_postgres = postgres_session.execute(query).fetchall()
        end_time = time.time()
        postgres_duration = end_time - start_time
        print("PostgreSQL Query Time:", postgres_duration, "seconds")
        postgres_session.commit()

    except Exception as e:
        print(f"An error occurred: {e}")
        postgres_session.rollback()
        
        
    # Execute and time the ClickHouse query
    try:
        start_time = time.time()
        clickhouse_session.execute(query).fetchall()
        clickhouse_duration = time.time() - start_time
        print("ClickHouse Query Time:", clickhouse_duration, "seconds")
    except Exception as e:
        print(f"ClickHouse error occurred: {e}")



## Example query

Query to calculate the total number of transactions and the average transaction amount from the transactions dataset.

In [22]:
query = '''SELECT 
    vendorid, 
    COUNT(*) AS trip_count
FROM 
    trips_taxi
GROUP BY 
    vendorid
ORDER BY 
    trip_count DESC;

'''

q0 = run_query(query)

q0

,vendorid,trip_count
0,2.0,155099
1,1.0,88329
2,0.0,6293
3,4.0,279


In [23]:
compare_query_runtime(query)

PostgreSQL Query Time: 0.012626171112060547 seconds
ClickHouse Query Time: 0.00546574592590332 seconds


# Questions


#### Q1. Write a query to find the number of trips, average passenger count, average trip distance, total fare amount, and total tip amount for each combination of vendor and rate code, where payment type is 1, sorted by trip count in descending order.


In [24]:
# Example usage (assuming postgres_session and clickhouse_session are already set up)
# TODO: YOUR CODE STARTS HERE
query = '''
SELECT
    vendorid,
    ratecodeid,
    COUNT(*) AS number_of_trips,
    AVG(passenger_count) AS avg_passenger_count,
    AVG(trip_distance) AS avg_trip_distance,
    SUM(fare_amount) AS total_fare_amount,
    SUM(tip_amount) AS total_tip_amount
FROM 
    trips_taxi
WHERE 
    payment_type = 1
GROUP BY 
    vendorid, ratecodeid
ORDER BY 
    number_of_trips DESC;
'''
# TODO: YOUR CODE ENDS HERE

q1 = run_query(query)

q1

,vendorid,ratecodeid,number_of_trips,avg_passenger_count,avg_trip_distance,total_fare_amount,total_tip_amount
0,2.0,1.0,105351,1.692447,2.651179,1.239073e+06,294868.519483
1,1.0,1.0,60450,1.168122,2.451267,7.075753e+05,160613.040134
2,2.0,2.0,2865,1.701222,16.800862,1.489800e+05,32408.459866
3,1.0,2.0,1326,1.308446,17.347964,6.895200e+04,14024.429993
4,2.0,5.0,845,1.506509,4.648367,4.855118e+04,6581.070010
5,2.0,3.0,269,1.773234,17.113903,1.839150e+04,3806.419996
6,1.0,5.0,224,1.272321,5.479464,1.327475e+04,2103.779992
7,4.0,1.0,186,1.043011,2.413763,2.163500e+03,550.909999
8,1.0,3.0,139,1.309353,16.890648,9.395000e+03,1939.859988
9,2.0,4.0,90,1.700000,22.852333,7.941500e+03,1258.530004


In [25]:
compare_query_runtime(query)

PostgreSQL Query Time: 0.039804697036743164 seconds
ClickHouse Query Time: 0.008043766021728516 seconds


#### Q2. Write a query to calculate the average fare amount and average tip amount for each vendor, considering only vendors with more than 1000 trips. The results should be ordered by average tip amount in descending order

In [26]:
# TODO: YOUR CODE STARTS HERE
query = '''
SELECT
    vendorid,
    COUNT(*) AS number_of_trips,
    AVG(fare_amount) AS avg_fare_amount,
    AVG(tip_amount) AS avg_tip_amount
FROM
    trips_taxi
GROUP BY
    vendorid
HAVING
    COUNT(*) > 1000
ORDER BY
    avg_tip_amount DESC;
'''
# TODO: YOUR CODE ENDS HERE
q2 = run_query(query)

q2

,vendorid,number_of_trips,avg_fare_amount,avg_tip_amount
0,2.0,155099,13.006355,2.184701
1,1.0,88329,12.619288,2.032444
2,0.0,6293,27.549450,1.079538


In [27]:
compare_query_runtime(query)

PostgreSQL Query Time: 0.012348651885986328 seconds
ClickHouse Query Time: 0.005954265594482422 seconds


#### Q3. Write a query to calculate the average fare amount for each pair of pickup and drop-off locations. The results should be ordered by average fare amount in descending order, and only the top 3 pairs with the highest average fare amount should be returned.

In [28]:
# TODO: YOUR CODE STARTS HERE
query = '''
SELECT
    pulocationid,
    dolocationid,
    AVG(fare_amount) AS avg_fare_amount
FROM
    trips_taxi
GROUP BY
    pulocationid, dolocationid
ORDER BY
    avg_fare_amount DESC
LIMIT 3;
'''
# TODO: YOUR CODE ENDS HERE
q3 = run_query(query)

q3

,pulocationid,dolocationid,avg_fare_amount
0,140,124,195.0
1,265,210,149.0
2,216,265,145.0


In [29]:
compare_query_runtime(query)

PostgreSQL Query Time: 0.028229951858520508 seconds
ClickHouse Query Time: 0.00668025016784668 seconds


#### Q4. Write a query to calculate the total fare amount and the percentage of fare amount for each payment type compared to the total fare amount for all trips. The results should be ordered by fare amount percentage in descending order.

In [30]:
# TODO: YOUR CODE STARTS HERE
query = '''
SELECT
    payment_type,
    SUM(fare_amount) AS total_fare_amount,
    ROUND(
        (SUM(fare_amount) * 100.0 / (SELECT SUM(fare_amount) FROM trips_taxi))::numeric,
        2
    ) AS fare_percentage
FROM
    trips_taxi
GROUP BY
    payment_type
ORDER BY
    fare_percentage DESC; 
'''
# TODO: YOUR CODE ENDS HERE
q4 = run_query(query)

q4

,payment_type,total_fare_amount,fare_percentage
0,1.0,2.268150e+06,68
1,2.0,8.523258e+05,25
2,0.0,1.733687e+05,5
3,4.0,1.313780e+03,0
4,3.0,1.378613e+04,0


In [31]:
compare_query_runtime(query)

PostgreSQL Query Time: 0.02009129524230957 seconds
ClickHouse Query Time: 0.021948814392089844 seconds


# ML

**Write a query for a taxi trip dataset that adds calculated features for each trip, excluding original timestamp columns but including the hour of the day and day of the week from the pickup time, and the trip duration in seconds. Use toHour, toDayOfWeek, and toUnixTimestamp functions in ClickHouse for these calculations.**

**Develop a regression model to predict taxi fare amounts using the dataset extracted from above query with numerical and categorical features. One-hot encode the categorical variables, train the model, and evaluate its performance by calculating the RMSE score.**

In [32]:
# TODO: YOUR CODE STARTS HERE
query = '''
SELECT
    vendorid,
    passenger_count,
    trip_distance,
    ratecodeid,
    store_and_fwd_flag,
    pulocationid,
    dolocationid,
    payment_type,
    fare_amount,
    extra,
    mta_tax,
    tip_amount,
    tolls_amount,
    improvement_surcharge,
    total_amount,
    congestion_surcharge,
    toHour(tpep_pickup_datetime) AS pickup_hour,
    toDayOfWeek(tpep_pickup_datetime) AS pickup_day_of_week,
    toUnixTimestamp(tpep_dropoff_datetime) - toUnixTimestamp(tpep_pickup_datetime) AS trip_duration_seconds
FROM
    trips_taxi
'''
# TODO: YOUR CODE ENDS HERE


In [33]:
data = run_query(query)

In [34]:
data.head()

,vendorid,passenger_count,trip_distance,ratecodeid,store_and_fwd_flag,pulocationid,dolocationid,payment_type,fare_amount,extra,mta_tax,tip_amount,tolls_amount,improvement_surcharge,total_amount,congestion_surcharge,pickup_hour,pickup_day_of_week,trip_duration_seconds
0,2.0,1.0,0.00,5.0,N,100,264,1.0,65.92,0.0,0.5,3.28,0.0,0.3,70.00,0.0,21,1,2
1,2.0,1.0,0.76,1.0,N,237,141,1.0,5.00,0.5,0.5,0.70,0.0,0.3,9.50,2.5,21,1,218
2,2.0,1.0,0.46,1.0,N,249,113,1.0,4.00,0.5,0.5,1.50,0.0,0.3,9.30,2.5,21,1,171
3,2.0,2.0,4.06,1.0,N,140,112,2.0,16.50,0.5,0.5,0.00,0.0,0.3,20.30,2.5,21,1,1188
4,2.0,1.0,3.96,1.0,N,43,234,1.0,15.00,0.5,0.5,2.82,0.0,0.3,21.62,2.5,21,1,1011


In [35]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

from sklearn.linear_model import LogisticRegression,LinearRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, accuracy_score ,confusion_matrix, roc_auc_score, mean_squared_error

In [36]:
# TODO: YOUR CODE STARTS HERE
# Drop fare_amount and total_amount to avoid data leakage
X = data.drop(columns=['fare_amount', 'total_amount'])
y = data['fare_amount']

# Categorical and numerical columns
categorical_columns = ['store_and_fwd_flag']
numerical_columns = [col for col in X.columns if col not in categorical_columns]

# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
# TODO: YOUR CODE ENDS HERE

In [37]:
# Training the model
# TODO: YOUR CODE STARTS HERE

# Transformers
numeric_transformer = Pipeline(steps=[('scaler', StandardScaler())])
categorical_transformer = Pipeline(steps=[('onehot', OneHotEncoder(handle_unknown='ignore'))])

# ColumnTransformer
preprocessor = ColumnTransformer(transformers=[
    ('num', numeric_transformer, numerical_columns),
    ('cat', categorical_transformer, categorical_columns)
])

# Full pipeline with Linear Regression
model = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('regressor', LinearRegression())
])

# Train
model.fit(X_train, y_train)

# TODO: YOUR CODE ENDS HERE



Pipeline(steps=[('preprocessor',
                 ColumnTransformer(transformers=[('num',
                                                  Pipeline(steps=[('scaler',
                                                                   StandardScaler())]),
                                                  ['vendorid',
                                                   'passenger_count',
                                                   'trip_distance',
                                                   'ratecodeid', 'pulocationid',
                                                   'dolocationid',
                                                   'payment_type', 'extra',
                                                   'mta_tax', 'tip_amount',
                                                   'tolls_amount',
                                                   'improvement_surcharge',
                                                   'congestion_surcharge',
                                                   'pickup_hour',
                                                   'pickup_day_of_week',
                                                   'trip_duration_seconds']),
                                                 ('cat',
                                                  Pipeline(steps=[('onehot',
                                                                   OneHotEncoder(handle_unknown='ignore'))]),
                                                  ['store_and_fwd_flag'])])),
                ('regressor', LinearRegression())])

In [38]:
# Predicting the Test set results
# TODO: YOUR CODE STARTS HERE

# Predict
y_pred = model.predict(X_test)

# RMSE
rmse = np.sqrt(mean_squared_error(y_test, y_pred))
print(f"RMSE: {rmse:.2f}")

# TODO: YOUR CODE ENDS HERE


RMSE: 8.31


In [40]:
y.describe()

count    250000.000000
mean         13.235776
std          12.308147
min         -75.000000
25%           6.500000
50%           9.500000
75%          15.000000
max         522.500000
Name: fare_amount, dtype: float64

The RMSE of 8.31 relative to a mean fare of $13.24 gives a 63% error rate, indicating moderate performance. Since fare amounts are non-linear, driven by distance, location, and rate codes. Linear Regression struggled to capture the full pricing patterns.